ci/Jenkinsfile.test-e2e.android

```
@@ -0,0 +1,277 @@
+      name: 'GIT_REF',
+      description: 'Git branch to checkout.',
+      branchFilter: 'origin/(.*)',
+      branch: '',
+      defaultValue: 'master',
+      quickFilterEnabled: false,
+      selectedValue: 'DEFAULT',
+      sortMode: 'ASCENDING_SMART',
+      tagFilter: '*',
+      type: 'PT_BRANCH'
+    )
+    string(
+      name: 'BUILD_SOURCE',
+      description: 'URL to APK or Jenkins build (pkg/*.apk). Required when BROWSERSTACK_APP_ID is empty.',
+      defaultValue: ''
+    )
+    string(
+      name: 'BROWSERSTACK_APP_ID',
+      description: 'Existing BrowserStack app identifier (lt://...). Leave empty to upload from BUILD_SOURCE.',
+      defaultValue: ''
+    )
+    string(
+      name: 'DEVICE_ID',
+      description: 'BrowserStack device id to run on (leave empty to use the environment default).',
+      defaultValue: ''
+    )
+    string(
+      name: 'PYTEST_ARGS',
+      description: 'Pytest flags (e.g. "-m smoke" or "-m critical") and other args.',
+      defaultValue: '-n=5 -m smoke tests'
+    )
+  }
+
+  options {
+    timestamps()
+    timeout(time: 120, unit: 'MINUTES')
+    buildDiscarder(logRotator(
+      daysToKeepStr: '30',
+      numToKeepStr: '30',
+      artifactNumToKeepStr: '30',
+    ))
+    disableRestartFromStage()
+  }
+
+  environment {
+    VIRTUAL_ENV = "${env.WORKSPACE_TMP}/venv-appium"
+    PYTHONUNBUFFERED = "1"
+    BUILD_SOURCE = "${params.BUILD_SOURCE}"
+    BROWSERSTACK_APP_ID_PARAM = "${params.BROWSERSTACK_APP_ID}"
+    PYTEST_ARGS = "${params.PYTEST_ARGS}"
+    DEVICE_ID = "${params.DEVICE_ID}"
+  }
+
+  stages {
+    stage('Checkout Reference') {
+      steps {
+        script {
+          def branch = params.GIT_REF ?: 'origin/master'
+          checkout([
+            $class: 'GitSCM',
+            branches: [[name: branch]],
+            doGenerateSubmoduleConfigurations: false,
+            extensions: scm.extensions,
+            submoduleCfg: [],
+            userRemoteConfigs: scm.userRemoteConfigs
+          ])
+          env.GIT_COMMIT = sh(script: 'git rev-parse HEAD', returnStdout: true).trim()
+          echo "Checked out ${branch} @ ${env.GIT_COMMIT}"
+        }
+      }
+    }
+
+    stage('Prep') {
+      steps {
+        script {
+          setBrowserStackProjectName()
+          updateGitHubStatus()
+        }
+      }
+    }
+
+    stage('Setup Python environment') {
+      steps {
+        dir('test/e2e_appium') {
+          sh """
+            set -euo pipefail
+            python3 -m venv ${VIRTUAL_ENV}
+            source ${VIRTUAL_ENV}/bin/activate
+            pip install --upgrade pip
+            pip install -r requirements.txt
+          """
+        }
+      }
+    }
+
+    stage('Acquire APK') {
+      steps {
+        script {
+          if (env.BROWSERSTACK_APP_ID_PARAM?.trim()) {
+            env.BROWSERSTACK_APP_ID = env.BROWSERSTACK_APP_ID_PARAM.trim()
+            echo "Using provided BrowserStack app id: ${env.BROWSERSTACK_APP_ID}"
+          } else {
```

This is why God invented `when` clauses for Jenkins stages:

```suggestion
      when { expression { env.BROWSERSTACK_APP_ID_PARAM?.trim() != '' } }
      steps {
        script {
```

---

Updated to use `when`